src/preprocess.inl.h

```diff
@@ -23,46 +23,46 @@
 #define FunctionReturnValue(function_index, local_return_index)  // Unnecessary, because `ReturnLocalStmt`.
 
 #define BeginFunction(function_index, function_name) \
-  if (!(function_index == ctx.opa_function_bodies.size())) { \
+  if (!(function_index == policy.function_bodies.size())) { \
     throw std::logic_error("Internal invariant failed."); \
   } \
-  ctx.opa_function_bodies.push_back( \
-    [&inl_opa_functions](CTX const& outer_ctx) -> typename CTX::ir_t { \
-    CTX ctx(typename CTX::ScopeDown(), outer_ctx); \
-    return stmtBlock(ctx, {
+  policy.function_bodies.push_back( \
+    [&inl_functions](POLICY const& policy) -> typename POLICY::code_t { \
+    return stmtBlock({
 #define EndFunction(function_index, function_name) }); });
 
-#define BeginBlock() stmtBlock(ctx, {
+#define BeginBlock() stmtBlock({
 #define EndBlock() }),
 
-#define ArrayAppendStmt(array, value) stmtArrayAppend(ctx, array, value),
-#define AssignIntStmt(value, target) stmtAssignInt(ctx, value, target),
-#define AssignVarOnceStmt(source, target) stmtAssignVarOnce(ctx, source, target),
-#define AssignVarStmt(source, target) stmtAssignVar(ctx, source, target),
-#define CallStmtBegin(func, target) stmtCall(ctx, func, {
+#define CallStmtBegin(func, target) stmtCall(func, {
 #define CallStmtPassArg(arg_index, arg_value) arg_value,
 #define CallStmtEnd(func, target) }, target),
-#define DotStmt(source, key, target) stmtDot(ctx, source, key, target),
-#define EqualStmt(a, b) stmtEqual(ctx, a, b),
-#define IsArrayStmt(array) stmtIsArray(ctx, array),
-#define IsDefinedStmt(source) stmtIsDefined(ctx, source),
-#define IsObjectStmt(source) stmtIsObject(ctx, source),
-#define IsUndefinedStmt(source) stmtIsUndefined(ctx, source),
-#define LenStmt(source, target) stmtLen(ctx, source, target),
-#define MakeArrayStmt(capacity, target) stmtMakeArray(ctx, capacity, target),
-#define MakeNullStmt(target) stmtMakeNull(ctx, target),
-#define MakeNumberIntStmt(value, target) stmtMakeNumberInt(ctx, value, target),
-#define MakeNumberRefStmt(index, target) stmtMakeNumberRef(ctx, index, target),
-#define MakeObjectStmt(target) stmtMakeObject(ctx, target),
-#define MakeSetStmt(target) stmtMakeSet(ctx),
-#define NotEqualStmt(a, b) stmtNotEqual(ctx, a, b),
-#define ObjectInsertOnceStmt(key, value, object) stmtObjectInsertOnce(ctx, key, value, object),
-#define ObjectInsertStmt(key, value, object) stmtObjectInsert(ctx, key, value, object),
-#define ObjectMergeStmt(a, b, target) stmtObjectMerge(ctx, a, b, target),
-#define ResetLocalStmt(target) stmtResetLocal(ctx, target),
-#define ResultSetAddStmt(value) stmtResultSetAdd(ctx, value),
-#define ReturnLocalStmt(source) stmtReturnLocalStmt(ctx, source),
-#define SetAddStmt(value, set) stmtSetAdd(ctx, value, set),
+
+#define ArrayAppendStmt(array, value) stmtArrayAppend(array, value),
```

Comment:
I'm sure this can be done cleaner, but it would require tweaking the DSL format.